forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/as/service/registrationrequest/RegistrationRequestFactory.java

```diff
@@ -0,0 +1,159 @@
+    ObjectMapper objectMapper;
+
+    @Autowired
+    public RegistrationRequestFactory(TppRegistrationService tppRegistrationService,
+                                      DirectorySoftwareStatementFactory softwareStatementFactory,
+                                      ObjectMapper objectMapper){
+        this.tppRegistrationService = tppRegistrationService;
+        this.softwareStatementFactory = softwareStatementFactory;
+        this.objectMapper = objectMapper;
+    }
+
+    public RegistrationRequest getRegistrationRequestFromJwt(String registrationRequestJwtSerialised)  throws DynamicClientRegistrationException {
+
+        try {
+            RegistrationRequestJWTClaims registrationRequestJWTClaims =
+                    getJwtClaimsSet(registrationRequestJwtSerialised, JWTClaimsOrigin.REGISTRATION_REQUEST_JWT);
+            String ssaJwtSerialised =
+                    registrationRequestJWTClaims.getRequiredStringClaim(RegistrationTppRequestClaims.SOFTWARE_STATEMENT);
+            RegistrationRequestJWTClaims ssaJwtClaims = getJwtClaimsSet(ssaJwtSerialised,
+                    JWTClaimsOrigin.SOFTWARE_STATEMENT_ASSERTION);
+            String issuer = ssaJwtClaims.getRequiredStringClaim(OpenBankingConstants.SSAClaims.ISSUER);
+
+            tppRegistrationService.validateSsaAgainstIssuingDirectoryJwksUri(ssaJwtSerialised, issuer);
+            DirectorySoftwareStatement regRequestSoftwareStatement =
+                    softwareStatementFactory.getSoftwareStatement(ssaJwtClaims);
+
+            String softwareJwkUri = regRequestSoftwareStatement.getSoftware_jwks_endpoint();
+            String softwareClientId =
+                    ssaJwtClaims.getRequiredStringClaim(OpenBankingConstants.SSAClaims.SOFTWARE_CLIENT_ID);
+            tppRegistrationService.verifyTPPRegistrationRequestSignature(registrationRequestJwtSerialised,
+                    softwareClientId, softwareJwkUri);
+            Map<String, Object> registrationRequestClaimsAsJson = registrationRequestJWTClaims.toJSONObject();
+            String registrationRequestJson = JSONObjectUtils.toJSONString(registrationRequestClaimsAsJson);
+
+            try {
+                RegistrationRequest request = objectMapper.readValue(registrationRequestJson, RegistrationRequest.class);
+                request.setJson(registrationRequestJson);
+                request.setDirectorySoftwareStatement(regRequestSoftwareStatement);
+                return request;
+            } catch (IOException ioe){
+
+                throw new DynamicClientRegistrationException("Could not map registration request jwt fields to " +
+                        "internal object " + ioe.getMessage(), DynamicClientRegistrationErrorType.INVALID_CLIENT_METADATA);
+            }
+        } catch (ParseException pe){
+            String errorMessage = "Could not parse the provided registration JWT";
+            log.debug("getRegistrationRequest() {}", errorMessage, pe);
```

Comment:
Could be better log.error or log.warn instead of log.debug? because contains the error message and exception